src/main/scheduler/scheduler.c

```diff
@@ -325,15 +343,62 @@ FAST_CODE void scheduler(void)
         }
 
         if (task->dynamicPriority > selectedTaskDynamicPriority) {
-            const bool taskCanBeChosenForScheduling =
-                (outsideRealtimeGuardInterval) ||
-                (task->taskAgeCycles > 1) ||
-                (task->staticPriority == TASK_PRIORITY_REALTIME);
+
+#ifdef PINIO_SCHEDULE_DEBUG
+            if ((task->staticPriority != TASK_PRIORITY_REALTIME) &&
+                (task->maxExecutionTime > remainingTime)) {
+                // Signal that a task could not be executed in the current loop
+                pinioSet(1,1);
+
+#if 0
+                if (task->taskAgeCycles > 1) {
+                    // Signal that a task will be executing late
+                    pinioSet(2,1);
+                } else {
+                    pinioSet(2,0);
+                }
+#endif
+            }
+#if 1
+            if (((task->maxExecutionTime > maxRemainingTime) || (task->taskAgeCycles > 1)) &&
+            	(remainingTime > averageRemainingTime)) {
+                // Signal that this task breaks determinism, but is being scheduled anyway
+                pinioSet(2,1);
+            } else {
+                pinioSet(2,0);
+            }
+#endif
+#endif //PINIO_SCHEDULE_DEBUG
+
+            // Determine if the current task is the best candidate so far for scheduling
+            bool taskCanBeChosenForScheduling;
+
+            /* If the task is realtime, let it run as soon as it's ready */
+            taskCanBeChosenForScheduling = (task->staticPriority == TASK_PRIORITY_REALTIME);
+
+            /* If the task will complete before the next realtime task must run, let it
+             * run.
+             */
+            taskCanBeChosenForScheduling |= task->maxExecutionTime <= remainingTime;
+
+            /* If a task is known to break determinism only schedule it if at least half the average
+             * remaining time remains in order to minimise the impact.
+             *
+             * Also consider the task age, as on startup a task may execute in less than the maxRemainingTime
+             * but subsequently the remaining time may never reach that maximum again.
+             */
+            taskCanBeChosenForScheduling |= (((task->maxExecutionTime > maxRemainingTime) || (task->taskAgeCycles > 1)) &&
+            		    					 (remainingTime > averageRemainingTime));
```

Comment:
As I've previously stated, using `maxExecutionTime` is not a good idea. A task can have peak timing events that are not indicative of their normal execution time. The common case is the initial execution the includes startup and initialization tasks, but we can't guarantee that other circumstances may occur in the task that set an unrepresentative max execution time. The underlying assumption that all the tasks should or can have run times that are "deterministic" or even consistent is naive and we can't base scheduler logic on this.

Reply:
It's not a question whether we "want" the task times to be consistent and "well behaved". The reality is that they're not at the moment and we can't guarantee that they will be any time soon. So basing scheduling logic on having consistent task times as a requirement is a non-starter.

Since we can't guarantee that there won't be spikes in task execution times, using a peak capturing maximum time doesn't produce acceptable results. Also as I mentioned earlier, the `maxExecutionTime` for each task is reset any time the CLI `tasks` command is run. The purpose of this value is to show the user a peak execution time since they last looked. So even more reason why this should not be the thing that determines if a task should be allowed to run.